### PR TITLE
Fix `cdist` backward calculation for `p=2`

### LINKED ISSUE #37154: torch.cdist produces nan gradients in Pytorch 1.5, but not Pytorch 1.4
## 🐛 Bug

torch.cdist produces nan gradients in Pytorch 1.5, but not Pytorch 1.4

## To Reproduce

1. Run the similarity example here: https://github.com/pytorch/vision/tree/master/references/similarity

2. On Pytorch 1.4, training proceeds normally with the loss decreasing each iteration

3. On Pytorch 1.5, the gradients become nan after the first iteration, which breaks training

4. I used torch.autograd.detect_anomaly to identify that torch.cdist is the problem (error trace below)

```python
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
~/Projects/vision/references/similarity/train.py in <module>
    171 if __name__ == '__main__':
    172     args = parse_args()
--> 173     main(args)

~/Projects/vision/references/similarity/train.py in main(args)
    132     for epoch in range(1, args.epochs + 1):
    133         print('Training...')
--> 134         train_epoch(model, optimizer, criterion, train_loader,
    135                     device, epoch, args.print_freq)
    136 

~/Projects/vision/references/similarity/train.py in train_epoch(model, optimizer, criterion, data_loader, device, epoch, print_freq)
     25         with torch.autograd.detect_anomaly():
     26             loss, frac_pos_triplets = criterion(embeddings, targets)
---> 27             loss.backward()
     28         optimizer.step()
     29 

~/miniconda3/envs/pytorch/lib/python3.8/site-packages/torch/tensor.py in backward(self, gradient, retain_graph, create_graph)
    196                 products. Defaults to ``False``.
    197         """
--> 198         torch.autograd.backward(self, gradient, retain_graph, create_graph)
    199 
    200     def register_hook(self, hook):

~/miniconda3/envs/pytorch/lib/python3.8/site-packages/torch/autograd/__init__.py in backward(tensors, grad_tensors, retain_graph, create_graph, grad_variables)
     96         retain_graph = create_graph
     97 
---> 98     Variable._execution_engine.run_backward(
     99         tensors, grad_tensors, retain_graph, create_graph,
    100         allow_unreachable=True)  # allow_unreachable flag

RuntimeError: Function 'SqrtBackward' returned nan values in its 0th output.
```

## Expected behavior

Training proceeds normally without nan gradients due to torch.cdist

## Environment

 - PyTorch Version (e.g., 1.0): Pytorch 1.5
 - OS (e.g., Linux): Ubuntu 18.04 LTS
 - How you installed PyTorch (`conda`, `pip`, source): conda install pytorch torchvision cudatoolkit=10.1 -c pytorch
 - Python version: 3.8.2
 - CUDA/cuDNN version: 10.1/7.6
 - GPU models and configuration: RTX 2080 Ti/Titan Xp (tried both)
 - Any other relevant information: I believe things work in Pytorch 1.4

cc @ezyang @gchanan @zou3519 @fmassa

### PR DESCRIPTION
**Pull request by emcastillo**
Closes #37154

Fixes a bug in `cdist` backward with `p=2`.
Under some circumstances, if the output has 0s, the gradient calculation of `sqrt` will be undefined. Leading to NaNs in the input gradients.

This PR defines a subgradient for this case.

A test is also added to verify this behavior, I was only able to reproduce it under certain shapes, so the shape is explicitly taken from #37154 example

### CODE REVIEW

**Comment by klicperajo** [comment]
Why do you use `1e-9` as epsilon and not something like `torch.finfo(result.dtype).min` (whatever the equivalent is in the C++ interface)?

**Comment by emcastillo** [comment]
`finfo` is implemented in the `csrc` layer which is not accessed by ATeN, and the `min` field is obtained with:

`std::numeric_limits<at::scalar_value_type<scalar_t>::type>::lowest());`
This requires knowing the result type in compile type as it is templatized, but in that function is dynamic so we have to resort to a type dispatching to do this calculation using `AT_DISPATCH_FLOATING__TYPES` but feels a bit overkill. (Please correct me if I am wrong, I find this a bit confusing and I am likely mistaken).

 `eps` values in ATeN codebase are usually manually set, so I just stick to that convention.

**Comment by klicperajo** [comment]
Yes, this might get complicated due to compilation. But the function needs to be compiled for every data type anyway, so imho it should still be possible.

The issue I have with using a fixed eps here is that the distances can be of varying scale. One set of cdists might be at the scale of 1e20 and the other at 1e-20. Both are perfectly valid and should not expect any loss of accuracy. A relative eps of 1e-9 would make sense in both cases (that would be around machine precision for float32), but imo not an absolute one.

**Comment by emcastillo** [comment]
I pushed the fix with your suggestion and verified that works. Thanks!

**Comment by klicperajo** [comment]
Nice, thank you! And apparently no underflows either. Great! :)

**Comment by vadimkantorov** [comment]
Will the computation for dynamic eps work correctly for torch.half / torch.bfloat16?

**Review by albanD — COMMENTED** [review]
Hey,
Thanks for the PR.
I'm not sure this is the right place to put this though. In general, we never change the result of the forward pass to make the backward stable. This would be surprising to users to do `cdist(a, a)` and not get 0.
The usual thing we do here is to modify the backward formula to add eps where needed. Or replace the value with special sub-gradients at given points.

Unfortunately, because in this case, the computation is split from the point of view of the autograd, it is expected that you get Nans. You will need to define a special backward formula for the overall op if you want to be able to provide a subgradient of 0 for that point.

Edit: related issue https://github.com/pytorch/pytorch/issues/37354

**Comment by emcastillo** [comment]
Hi @albanD

I thought about doing that at first because it seemed the most formally correct solution,
However, since the issue is the sqrt gradient, we end up needing to define either a new operation for the gradient of sqrt to be used only in cdist, or we need to write a complete backward pass Template. Also masking the zeros lead to issues in performance as @ngimel pointed out so I was not sure of which path to take :).

I don't have any objections about this so If you guys think that this approach is better I will change the PR accordingly!

**Comment by emcastillo** [comment]
@vadimkantorov I think this fix does not work for float16 since the min value seems to be set to.
```
>>> torch.finfo(torch.float16).min
-65504.0
```

**Comment by gchanan** [comment]
@emcastillo you probably want something like `eps` or `tiny`, instead of `min`, see the documentation:
https://pytorch.org/docs/master/type_info.html?highlight=finfo#torch.torch.finfo

**Comment by albanD** [comment]
Actually I would argue the problem is not with sqrt. The gradient of sqrt at 0 is "inf".
The problem is that "norm" as a function has a proper gradient defined at 0. But if you use the formula used here to compute here, because you're at a point where one of the elementary function is not continuously differentiable (sqrt at 0), you cannot use the chain rule in theory (or at least it can give you arbitrarily wrong values). This is what happens here :/ 

So the solution here would be not to use the chain rule and write the derivative for norm directly. But as you mention, we still have to mask out the values at 0 because we don't have a general formula that handles it properly.


Even though I agree that this would be more efficient than the masking, I am still not sure we want to start modifying the forward values just to avoid problematic points in the backward: it doesn't feel like the right solution and we never do it in the current codebase. So I don't think we should start (this is my personal opinion, so potentially open to discussion if other people think otherwise).

I feel like the right solution here would be to
- rename `euclidean_dist_out` -> `_euclidean_dist`: this is not an `_out` function, not sure why it's named that way
- add en entry to `native_functions.yaml` for `_euclidean_dist`.
- add en entry to `derivatives.yaml` for `_euclidean_dist` setting the derivative as `_euclidean_dist_backward`.
- add `_euclidean_dist_backward` to `tools/autograd/templates/Functions.cpp` where you implement the backward of the overall function and add the special case for 0 (note that you control the whole formula here so you can instead add an eps before dividing by the norm to avoid both nans and the masking).

**Comment by ngimel** [comment]
> @emcastillo you probably want something like `eps` or `tiny`, instead of `min`, see the documentation:
> https://pytorch.org/docs/master/type_info.html?highlight=finfo#torch.torch.finfo

For floating point numeric_limits::min corresponds to finfo.tiny, so that it's all logical and not confusing at all.

**Comment by emcastillo** [comment]
Hi,
I am fine with changing the PR to take @albanD's approach.
@ngimel what do you think?

Thank you

**Comment by ngimel** [comment]
@emcastillo sure, go ahead.

**Comment by emcastillo** [comment]
Thanks!

**Comment by emcastillo** [comment]
I would appreciate some help here.
I am trying to define a `_euclidean_dist_backward` in `tools/autograd/templates/Functions.cpp`

In this definition, I have to implement or call all the gradient definitions of all the operations that composed this cdist. However in cases such as `matmul` I don't find any backward definition for it that I can call from this code, and when calculating the backward of `clamp_min_` it requires a `min` parameter that I am not sure how and where to retrieve.

**Comment by emcastillo** [comment]
I did the following in order to simplify the implementation.

I broke down the original `euclidean_dist` function in two functions
`_euclidean_dist_squared` which just returns the squared distances as its name says.
and then `_sqrt_euclidean_dist` to do the square root of the squared euclidean distance.
Then I just defined a custom backward pass for `_sqrt_euclidean_dist`.

The reason why I did this, was because for intermediate steps of the composite operation such as `matrix_mult` it is pretty burdensome to define the gradient calculation as there is lots of cases in the forward pass. 

Also, a question here, If we were to define a single backward function for all the operations, how do we retrieve the results of the intermediate functions in the custom-defined backward function? I couldn't find an easy way for this.

File: test/test_autograd.py [hunk 1]
```diff
@@ -5304,6 +5303,21 @@ def _test_euclidean_large_cdist(sizex, sizey=None):
...
+            x = torch.randn(sizex, device=device, dtype=torch.float)
+            dist_grad = torch.randn((1, 27, 27), device=device, dtype=torch.float)
+            y = x.clone()
+            eps = 1e-6
+            x.requires_grad = True
+            d = torch.cdist(x, y)
+            d.backward(dist_grad)
+            # Check that the backward passs does not contain invalid 
+            # values such as nan or inf
+            assert not torch.isnan(x.grad).any() and not torch.isinf(x.grad).any()
```
Comment (ngimel) [inline test/test_autograd.py:5320]: `assert torch.isfinite(x.grad).all()`

File: tools/autograd/templates/Functions.cpp [hunk 2]
```diff
@@ -90,6 +90,13 @@ int64_t _safe_size(IntArrayRef sizes, IntArrayRef dim) {
   return size;
 }
 
+Tensor sqrt_euclidean_dist_backward(const Tensor & grad, const Tensor & self) {
+  // handle case at 0 where we return a subgradient containing 0
+  Tensor result = grad / (2 * self);
+  result.masked_fill_(self == 0, 0);
+  return result;
```
Comment (ngimel) [inline tools/autograd/templates/Functions.cpp:97]: @albanD suggested the following, which should be faster than masked_fill, but we did not benchmark. 
```
Tensor  scale_v;
// If norm is very small, self is also. So the result will be very small anyway
scale_v = grad / (2 * self.clamp_min(eps));
return self * scale_v;
```

**Review by albanD — COMMENTED** [review]
Following the comments below, I don't think this split is the way to go.

Check the script below, it contains the manual backward for the whole cdist. It will be faster than the autograd one and much more memory efficient. It only needs the inputs and outputs so no issue with intermediary results.

Now for the issue at hand. We can definitely do masking based on result's value there.
But maybe we can do better by just adding an epsilon as Natalia mentioned if it gets the exact same result.
Not sure it is possible though (as the example we discussed was for norm for which we knew that the norm can only be 0 if every element is 0, this is not true here anymore, especially for the second term that involves the matmuls).

```python
import torch
from torch import autograd
from torch.autograd.functional import jacobian

autograd.set_detect_anomaly(True)

B, P, R, M = 3, 1, 2, 4
flag = "use_mm_for_euclid_dist"

x1 = torch.rand(B, P, M, requires_grad=True)
x2 = torch.rand(B, R, M, requires_grad=True)

out = torch.cdist(x1, x2, compute_mode=flag)
jac = jacobian(lambda x1, x2: torch.cdist(x1, x2, compute_mode=flag), (x1, x2))

class CustomCdist(autograd.Function):
    @staticmethod
    def forward(ctx, x1, x2):
        # Taken from cpp code
        x1_norm = x1.pow(2).sum(-1, keepdim=True);
        x1_pad = torch.ones_like(x1_norm);
        x2_norm = x2.pow(2).sum(-1, keepdim=True);
        x2_pad = torch.ones_like(x2_norm);
        x1_ = torch.cat([x1.mul(-2), x1_norm, x1_pad], -1);
        x2_ = torch.cat([x2, x2_pad, x2_norm], -1);
        result = x1_.matmul(x2_.transpose(-2, -1));
        result.clamp_min_(0).sqrt_();
        ctx.save_for_backward(x1, x2, result)
        return result

    @staticmethod
    def backward(ctx, gres):
        x1, x2, result = ctx.saved_tensors
        ratio = gres / result # needs stabilization when entries in result are 0
        gx1 = (x1 * ratio.sum(-1, keepdim=True) - ratio.matmul(x2))
        gx2 = (x2 * ratio.sum(-2, keepdim=False).unsqueeze(-1) - ratio.transpose(-2, -1).matmul(x1))
        return gx1, gx2

custom_out = CustomCdist.apply(x1, x2)
custom_out.sum().backward()

print("same forward?", (out - custom_out).abs().max())

custom_jac = jacobian(lambda x1, x2: CustomCdist.apply(x1, x2), (x1, x2))

for t, custom_t in zip(jac, custom_jac):
    print("same jac?", (t - custom_t).abs().max())

print("Grad check: ", autograd.gradcheck(CustomCdist.apply, (x1.double(), x2.double())))
```

File: tools/autograd/templates/Functions.cpp (hunk 2 above)
Comment (albanD) [reply tools/autograd/templates/Functions.cpp:97]: Note that this was for norm. Won't apply as-is here because we don't have `self` here but `x1 - x2` :/

File: tools/autograd/templates/Functions.cpp [hunk 3]
```diff
@@ -90,6 +90,13 @@ int64_t _safe_size(IntArrayRef sizes, IntArrayRef dim) {
   return size;
 }
 
+Tensor sqrt_euclidean_dist_backward(const Tensor & grad, const Tensor & self) {
+  // handle case at 0 where we return a subgradient containing 0
+  Tensor result = grad / (2 * self);
+  result.masked_fill_(self == 0, 0);
```
[the remark this answers is not in the document]
Comment (emcastillo) [reply tools/autograd/templates/Functions.cpp:96]: Thanks for the snippet!
Let me try tomorrow morning 😅

But the way, just out of curiosity, how do we access the context in ATen?
I am a total newbie with this so sorry again for keep asking for guidance.

[the remark this answers is not in the document]
Comment (albanD) [reply tools/autograd/templates/Functions.cpp:96]: When converting the custom Function above in aten, you don't have a context, you make purely function version of these. You will just take as input (grad, x1, x2, result) and define it this way in derivatives.yaml.
See [here](https://github.com/pytorch/pytorch/blob/149b468ce288c9769bb2b4b535141876dbfec10f/tools/autograd/derivatives.yaml#L517-L518) for example. `grad` and `result` are automagically given to you. and if you want to reuse arguments, you can simply use their name.

Basically reusing the logic you have right now for your custom sqrt. But now make a backward for the whole function.

[the remark this answers is not in the document]
Comment (emcastillo) [reply tools/autograd/templates/Functions.cpp]: Yup, I got that part! 😃
My question was about accessing a partial result of the forward computation, for example inside the backward function, I want to access the result of matmul in the forward pass.
In Python there is the context object, but I didn't see a similar thing in ATen so far.

**Comment by emcastillo** [comment]
Changed the function with @albanD suggestions, thanks a lot!

File: tools/autograd/derivatives.yaml [hunk 4]
```diff
@@ -669,6 +669,10 @@
   self: not_implemented("_pdist_backward")
   pdist: not_implemented("_pdist_backward")
 
+- name: _euclidean_dist(Tensor x1, Tensor x2) -> Tensor
+  x1: _euclidean_dist_backward(grad, x1, x2, 0, result)
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:673]: this can be slightly simpler if you use a single function to do both: defining it like [this](https://github.com/pytorch/pytorch/blob/4ed790d74223ff3af95bbf8585cb67664805b6b9/tools/autograd/derivatives.yaml#L193-L194) as `x1, x2: xxx` and making your function return a tuple like [this](https://github.com/pytorch/pytorch/blob/4ed790d74223ff3af95bbf8585cb67664805b6b9/tools/autograd/templates/Functions.cpp#L1626-L1631).

**Review by albanD — COMMENTED** [review]
Thanks!
This looks much better I think.
Would you have time to make the change to combine them in a single function? Or shall we do that as a follow up?

**Comment by ngimel** [comment]
@alband is this ok to be landed? Can you do it?

**Comment by emcastillo** [comment]
Let me unify the derivatives step later today 😅

**Comment by emcastillo** [comment]
Rebased, PTAL

**Comment by albanD** [comment]
Yes this is good for landing.

The extra comment can be removed but I'll do that in a follow up.
Thanks for taking the time!

**Comment by KevinMusgrave** [comment]
In which version was this fixed applied?
